File: 3d/coord64/C/benchmark.c
#define _POSIX_C_SOURCE 199309L
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "kd.h"

double get_time() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}

int main() {
    kd_tree tree = kd_create();
    srand(42);

    long long (*boxes)[6] = malloc(1000000 * sizeof(*boxes));

    printf("Inserting 1000000 boxes...\n");
    double start = get_time();
    for (int i = 0; i < 1000000; i++) {
        long long x1 = ((long long)rand() * rand()) % 10000000000LL;
        long long y1 = ((long long)rand() * rand()) % 10000000000LL;
        long long z1 = ((long long)rand() * rand()) % 10000000000LL;
        long long x2 = x1 + (rand() % 100) + 1;
        long long y2 = y1 + (rand() % 100) + 1;
        long long z2 = z1 + (rand() % 100) + 1;
        
        boxes[i][0] = x1; 
	boxes[i][1] = y1; 
	boxes[i][2] = z1;
	
	boxes[i][3] = x2; 
	boxes[i][4] = y2;
	boxes[i][5] = z2;
        kd_insert(tree, (void *)(long)(i + 1), boxes[i], NULL);
    }
    printf("Insertion took %fs\n", get_time() - start);

    long long search_area[6] = {0, 0, 0, 5000000000LL, 5000000000LL, 5000000000LL};
    printf("Searching in area...\n");
    start = get_time();
    kd_gen gen = kd_start(tree, search_area);
    char *data;
    long long found_size[6];
    int found_count = 0;
    
    while (kd_next(gen, &data, found_size) == KD_OK) {
        found_count++;
    }
    kd_finish(gen);
    printf("Found %d boxes in search area. Search took %fs\n", found_count, get_time() - start);

    int tries, deleted;
    printf("Deleting 1000 items...\n");
    start = get_time();
    for (int i = 0; i < 1000; i++) {
        kd_really_delete(tree, (void *)(long)(i + 1), boxes[i], &tries, &deleted);
    }
    printf("Deletion took %fs\n", get_time() - start);

    free(boxes);
    return 0;
}

File: 3d/coord64/julia/benchmark.jl
using KDTree3D
using Random
import KDTree3D: insert!, search, hard_delete!

function benchmark()
    tree = Tree{Int}()
    Random.seed!(42)
    
    n_boxes = 1_000_000
    boxes = Vector{Box}(undef, n_boxes)
    
    println("Inserting $n_boxes boxes...")
    start = time()
    for i in 1:n_boxes
        x1 = rand(0:9_999_999_999)
        y1 = rand(0:9_999_999_999)
        z1 = rand(0:9_999_999_999)
        x2 = x1 + rand(1:100)
        y2 = y1 + rand(1:100)
        z2 = z1 + rand(1:100)
        
        boxes[i] = (x1, y1, z1, x2, y2, z2)
        insert!(tree, i, boxes[i])
    end
    println("Insertion took $(time() - start)s")
    
    search_area = (0, 0, 0, 5_000_000_000, 5_000_000_000, 5_000_000_000)
    println("Searching in area $search_area...")
    start = time()
    results = search(tree, search_area)
    println("Found $(length(results)) boxes in search area. Search took $(time() - start)s")
    
    println("Deleting 1000 items...")
    start = time()
    for i in 1:1000
        hard_delete!(tree, i, boxes[i])
    end
    println("Deletion took $(time() - start)s")
    println("Done.")
end

if abspath(PROGRAM_FILE) == @__FILE__
    benchmark()
end
